src/GitTfs.VsCommon/TfsHelper.Vs2017Base.cs

```diff
@@ -0,0 +1,201 @@
+            : base(bridge, container)
+        {
+            myMajorVersion = majorVersion;
+            myVisualStudioInstallationPath = GetVsInstallDir();
+
+            myAssemblySearchPaths = new List<string>();
+            if (!string.IsNullOrEmpty(myVisualStudioInstallationPath))
+            {
+                myAssemblySearchPaths.Add(Path.Combine(myVisualStudioInstallationPath, myPrivateAssembliesFolder));
+            }
+        }
+
+        protected override bool HasWorkItems(Changeset changeset)
+        {
+            return Retry.Do(() => changeset.AssociatedWorkItems.Length > 0);
+        }
+
+        /// <summary>
+        /// Enumerates the list of installed VS instances and returns the first one
+        /// matching <see cref="MajorVersion"/>. Right now there is no way for the user to influence
+        /// which version to choose if multiple installed version have the same major,
+        /// e.g. VS2017 installed as Enterprise and Professional.
+        /// </summary>
+        /// <returns>
+        /// Path to the top level directory of the Visual studio installation directory,
+        /// e.g. <c>C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise</c>
+        /// </returns>
+        protected string GetVsInstallDir()
+        {
+            if (myVisualStudioInstallationPath != null)
+                return myVisualStudioInstallationPath;
+
+            var setupConfiguration = (ISetupConfiguration2)GetSetupConfiguration();
+            IEnumSetupInstances instancesEnumerator = setupConfiguration.EnumAllInstances();
+
+            int fetched;
+            var instances = new ISetupInstance[1];
+            while (true)
+            {
+                instancesEnumerator.Next(1, instances, out fetched);
+                if (fetched <= 0)
+                    break;
+
+                var instance = (ISetupInstance2)instances[0];
+                if (!Version.TryParse(instance.GetInstallationVersion(), out Version version))
+                {
+                    Trace.TraceError("Failed to retrieve version. Skipping VS instance.");
+                    continue;
+                }
+
+                if (version.Major != myMajorVersion)
+                {
+                    continue;
+                }
+
+                if (myVisualStudioInstallationPath != null)
+                {
+                    Trace.TraceWarning("Already found a Visual Studio version. Ignoring version at {0}", instance.GetInstallationPath());
+                    continue;
+                }
+
+                var state = instance.GetState();
+                if (state.HasFlag(InstanceState.Local) && state.HasFlag(InstanceState.Registered))
+                {
+                    myVisualStudioInstallationPath = instance.GetInstallationPath();
+                    Trace.TraceInformation("Found matching Visual Studio version at {0}", myVisualStudioInstallationPath);
```

Comment:
This currently is printed to often, e.g. for each `tfs-remote` in the .git/config file.
The underlying issue is that the GitTfs.Vs2017\TfsHelper.cs is created multiple times and therefore its base class (this file) is constructed multiple times.

Would it be possible to make this class a singleton? That whould mean to configure it on the DI container as such, but
I am little nervous about the consequences why this wasn't done before already.

Having it recreated for each git-tfs-remote would also mean we do this detection each and every time again ;-(

Reply:
Yes, I don't know the consequences either 😢 
Perhaps storing the path in a static field could be a not so bad solution...

Reply:
I had exactly what you proposed already here - or at least a common static state which should be shared over all instance,, but as this is a base class for theVS2017 TfsHelper and the (yet to come) VS2019 TfsHelper I figured it is not practical.
It would be a possibility to do it like here https://stackoverflow.com/a/32107465, e.g. make the state depend on the concrete type of the class.
 
But then I figured I should ask here first on your thoughts about why the TfsHelper instances are not made a singleton in the structuremap DI container.

I did make the modifcations to make it singleton locally  and used to import a few changesets, but then - because I was unsure e.g. about sharing between threads and other implications - didn't follow through with it.

Reply:
>It would be a possibility to do it like here https://stackoverflow.com/a/32107465, e.g. make the state depend on the concrete type of the class, but this also feels a little clumsy.

You could do that but instead of the class name, use the `majorVersion` as the key.

I would go with this idea because I also don't know the consequences to put `TfsHelper` instances as singleton...

Reply:
Yes, of course that would be better. 
I will try to implement this and add it to this PR, otherwise you get a lot of console outputs in case you interact with multiple TFS remotes. Currently, one line is written for each TFS branch we are dealing with, which is quite noisy!

Reply:
@siprbaum I don't recall any particular reason for instantiating the plugin multiple times. IIRC it's that way because it's the default in structure map. I took a quick look through the code and don't see any state that would be problematic to have as a singleton. I think a singleton would be better than storing things in a static var on an abstract class. :+1: on marking the tfs helper as a singleton.

Reply:
@spraints OK, I will look into it, making it a singleton if you think it is clearer. (I think this too, but currently it feels riskyier,
because this would also affect the working VS2015 support if there are unforseen sideeffects, which I tried to avoid till now)

Reply:
Added a commit to make the TfsHelper a singleton